[Workflow management UI changes 1] Implemented task chooser modal

kaedroho opened the pull request:
This adds a new task chooser modal. The modal allows the creation of new tasks or choosing an existing one.

This PR doesn't include:

 - Converting tasks to use plain Django forms
 - Admin listing fields for tasks
 - Fix for the checkbox widget styling

These will be included in future PRs.

Separated from: https://github.com/wagtail/wagtail/pull/6157

kaedroho commented on wagtail/admin/views/workflows.py:
```diff
@@ -62,7 +70,13 @@ def get_edit_handler(self):
-        return self.get_edit_handler().get_form_class()
+        form_class = self.get_edit_handler().get_form_class()
+
+        # TODO Unhack
```
I've unhacked this in my other PR

kaedroho commented on wagtail/admin/views/workflows.py:
```diff
@@ -99,7 +113,13 @@ def get_edit_handler(self):
-        return self.get_edit_handler().get_form_class()
+        form_class = self.get_edit_handler().get_form_class()
+
+        # TODO Unhack
```
Ditto

kaedroho commented on wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html:
```diff
@@ -0,0 +1,62 @@
+                    {% trans "Name" %}
+                </th>
+                {% if searchform.specific_task_model_selected %}
+                    {# TODO #}
```
Going to leave this for a separate PR. We need to display some custom fields here and on the task index

zerolab commented on wagtail/admin/forms/workflows.py:
```diff
@@ -1,9 +1,77 @@
+
+
+def model_name(model):
+    return model._meta.app_label + '.' + model.__name__
```
could this live in `core.utils`? especially that it contains `resolve_model_string` which is the opposite of this, in a way

zerolab commented on wagtail/admin/forms/workflows.py:
```diff
@@ -1,9 +1,77 @@
+        # Save a mapping of task_type values back to the model that we can reference later
+        self.task_type_choices = {
+            model_name(model): model
+            for model, _ in task_type_choices
```
can you rework this in light of @gasman's comment - https://github.com/wagtail/wagtail/pull/6163#issuecomment-647579099:

> for example, using _ as the return value of get_or_create is liable to clash with the _ alias commonly used for gettext

zerolab commented on wagtail/admin/templates/wagtailadmin/workflows/task_chooser/chooser.html:
```diff
@@ -0,0 +1,38 @@
+            </ul>
+        </form>
+        <div id="search-results" class="listing tasks">
+            {% include "./includes/results.html" %}
```
All other chooser templates use full paths, let's be consistent. There are relative includes above too

zerolab commented on wagtail/admin/templates/wagtailadmin/workflows/task_chooser/includes/results.html:
```diff
@@ -0,0 +1,62 @@
+                        <div class="title-wrapper"><a href="{% url 'wagtailadmin_workflows:task_chosen' task.id %}" class="task-choice">{{ task.name }}</a></div>
+                    </td>
+                    {% if searchform.specific_task_model_selected %}
+                        {# TODO #}
```
Is this done in any of the follow-up PRs?
Can't see it in #6171 or #6172

kaedroho replied:
Nope, this hasn't been implemented yet

zerolab replied:
👍

zerolab commented on wagtail/admin/views/workflows.py:
```diff
@@ -452,3 +472,146 @@ def enable_task(request, pk):
+        'step': 'chooser',
+        'error_label': _("Server Error"),
+        'error_message': _("Report this error to your webmaster with the following information:"),
+        'tag_autocomplete_url': reverse('wagtailadmin_tag_autocomplete'),
```
This is not used anywhere, is it needed?

zerolab commented on wagtail/admin/views/workflows.py:
```diff
@@ -452,3 +472,146 @@ def enable_task(request, pk):
+    # TODO
+    #  # allow hooks to modify the queryset
+    # for hook in hooks.get_hooks('construct_document_chooser_queryset'):
+    #     documents = hook(documents, request)
```
Looks like a remnant from copying the document_chooser method. Would be good to cleanup

kaedroho commented:
Thanks for the review @zerolab. I've addressed your feedback in: https://github.com/wagtail/wagtail/pull/6169/commits/e2359c914f23bbfaa1446f42773452b652c5b661
